### PR TITLE
Modifies the symbolic suffixes enclosing them into brackets and sympl…

### PR DESCRIPTION
**Pull request by LoCorVin**
…ifying them with the z3 simplification algorithm. Should solve #303

### CODE REVIEW

**Comment by norhh** [comment]
Hi @LoCorVin , now it would be quite helpful to understand what is happening while going through graphs etc thanks to your PR.
Currently laser underwent lot of changes, can you change the code based on that ,all the instructions are currently in instructions.py .

**Comment by LoCorVin** [comment]
Hope this is sufficient :)

File: mythril/laser/ethereum/instructions.py [hunk 1]
```diff
@@ -382,24 +382,29 @@ def calldatacopy_(self, global_state):
...
+        size_sym = False
         try:
             size = util.get_concrete_int(op2)
+            size_sym = True
             # FIXME: broad exception catch
         except:
             logging.debug("Unsupported symbolic size in CALLDATACOPY")
+            size = simplify(op2)
+
+        if dstart_sym or size_sym:
```
Comment (norhh) [inline mythril/laser/ethereum/instructions.py]: the True and False should be swapped right? .If we get concrete ints for both then we can do the copying onto memory

File: mythril/laser/ethereum/svm.py [hunk 2]
```diff
@@ -575,20 +575,25 @@ def _sym_exec(self, gblState):
...
 
+                size_sym = False
                 try:
                     size = helper.get_concrete_int(op2)
+                    size_sym = True
                 except:
                     logging.debug("Unsupported symbolic size in CALLDATACOPY")
+                    size = simplify(op2)
+
+                if dstart_sym or size_sym:
```
[the remark this answers is not in the document]
Comment (LoCorVin) [reply mythril/laser/ethereum/svm.py]: Yes, your right, a really obvious mistake.

**Comment by norhh** [comment]
hey @LoCorVin can you rerun the builds by merging with the master(as build failed due to some ethereum version issues which was corrected in master) Also I see some huge changes, can you mention what these changes will do, that might help a lot.

File: mythril/solidnotary/z3utility.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+def are_satisfiable(constraints):
+    s = Solver()
+    for c in constraints:
+        s.add(c)
+    return s.check() == sat
+
+def simplify_constraints(constraints): # Todo simplification of the sum of constraints
+    simp_const = []
+    for const in constraints:
+        simp_const.append(simplify(const))
```
Comment (norhh) [inline mythril/solidnotary/z3utility.py]: a list comprehensive would be quick here,
just 
simp_const = [simplify(const] for const in constraints] 
or something like that

Comment (LoCorVin) [reply mythril/solidnotary/z3utility.py]: sorry, these are changes from my master branch, I must have done something horribly wrong when merging in your main master or when pushing. I will figure out how to revert to the last accetable point.

Comment (norhh) [reply mythril/solidnotary/z3utility.py]: After reverting can you merge your branch with the current master of mythril else build might fail

**Pull request merged**